crates/wdl-analysis/src/types.rs

```diff
@@ -911,6 +1070,137 @@ impl Coercible for StructType {
     }
 }
 
+/// Represents the type of an enum.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EnumType {
+    /// The name of the enum.
+    name: Arc<String>,
+    /// The common coerced type computed from all variant values.
+    inner_value_type: Type,
+    /// The variants.
+    variants: Arc<Vec<String>>,
+}
+
+impl EnumType {
+    /// Constructs a new enum type with a known coerced type.
+    ///
+    /// Validates that all variant types are coercible to the provided type.
+    ///
+    /// Returns an error if any variant cannot be coerced.
+    pub fn new(
+        enum_name: impl Into<String>,
+        enum_span: Span,
+        explicit_inner_type: Type,
+        variants: Vec<(String, Type)>,
+        variant_spans: &[Span],
+    ) -> Result<Self, Diagnostic> {
+        assert_eq!(variants.len(), variant_spans.len());
+        let enum_name = enum_name.into();
+
+        // Validate that all variant types are coercible to the value type
+        for (variant_idx, (variant_name, variant_type)) in variants.iter().enumerate() {
+            if !variant_type.is_coercible_to(&explicit_inner_type) {
+                return Err(crate::diagnostics::enum_variant_does_not_coerce_to_type(
+                    &enum_name,
+                    enum_span,
+                    variant_name,
+                    variant_spans[variant_idx],
+                    &explicit_inner_type,
+                    variant_type,
+                ));
+            }
+        }
+
+        let variants = variants.into_iter().map(|(name, _)| name).collect();
```

Comment:
Nit: could probably just have a `Vec::with_capacity(variants.len())` that we push the name into with a single iteration over `variants.into_iter().enumerate()` above (doesn't really matter though).